File: kt/Minsk/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/binding/BoundScope.kt
package dev.phytolizer.minsk.analysis.binding

import dev.phytolizer.minsk.analysis.VariableSymbol

internal class BoundScope(val parent: BoundScope?) {
    val _variables = mutableMapOf<String, VariableSymbol>()

    fun tryDeclare(variable: VariableSymbol): Boolean {
        return if (_variables.containsKey(variable.name)) {
            false
        } else {
            _variables[variable.name] = variable
            true
        }
    }

    fun tryLookup(name: String): VariableSymbol? {
        val variable = _variables[name]
        if (variable != null) {
            return variable
        }

        if (parent == null) {
            return null
        }

        return parent.tryLookup(name)
    }

    fun declaredVariables() = _variables.values.toList()
}

File: cs/Minsk/CodeAnalysis/Binding/BoundScope.cs
﻿using System.Collections.Immutable;

namespace Minsk.CodeAnalysis.Binding;

internal sealed class BoundScope
{
    public BoundScope? Parent { get; }
    private readonly Dictionary<string, VariableSymbol> _variables = new();

    public BoundScope(BoundScope? parent)
    {
        Parent = parent;
    }

    public bool TryDeclare(VariableSymbol variable)
    {
        if (_variables.ContainsKey(variable.Name))
        {
            return false;
        }

        _variables.Add(variable.Name, variable);
        return true;
    }

    public bool TryLookup(string name, out VariableSymbol variable)
    {
        if (_variables.TryGetValue(name, out variable!))
        {
            return true;
        }

        if (Parent == null)
        {
            return false;
        }

        return Parent.TryLookup(name, out variable);
    }

    public ImmutableArray<VariableSymbol> GetDeclaredVariables() => _variables.Values.ToImmutableArray();
}
